src/Illuminate/Foundation/Exceptions/Handler.php

```diff
@@ -486,9 +488,11 @@ protected function renderExceptionResponse($request, Throwable $e)
      */
     protected function unauthenticated($request, AuthenticationException $exception)
     {
-        return $this->shouldReturnJson($request, $exception)
+        $redirectTo = $exception->redirectTo() ?? (Route::has('login') ? route('login') : null);
+
+        return $this->shouldReturnJson($request, $exception) || ! $redirectTo
                     ? response()->json(['message' => $exception->getMessage()], 401)
-                    : redirect()->guest($exception->redirectTo() ?? route('login'));
+                    : redirect()->guest($redirectTo);
```

Comment:
Correct me if I'm wrong, but with this logic it is currently possible for a route to return a JSON response when it's not expected to, if there is no `$redirectTo` target. Is this correct behaviour?

Reply:
Yes, that's exactly the point of this change.

I am arguing that receiving `500 | Server error` or `Route [login] not defined.` is wrong (and is not useful) if the actual issue is that you're not authenticated and returning the JSON message is the most reasonable fallback in that case.